# Introduction to Next Word Prediction and GPT-2 Fine-tuning

Next word prediction is a fundamental task in natural language processing that serves as the foundation for training large language models like GPT-2. This technique works effectively for several reasons:

1. **It captures contextual relationships:** By predicting the next word based on previous words, the model learns to understand context and semantic relationships in language.
2. **Self-supervised learning:** It allows for unsupervised pre-training on vast amounts of text data without the need for labeled datasets.
3. **Generalization:** The skills learned through next word prediction transfer well to various downstream NLP tasks.



![Screenshot](/images/1.png)

In this project, we're **fine-tuning GPT-2 on a cricket commentary dataset** to generate cricket-specific text. Fine-tuning adapts the pre-trained model to our specific domain, allowing it to generate more relevant and accurate cricket commentary.



### Data Preparation

We start by loading and cleaning our **cricket commentary data**. 

Cleaning involves 
- removing punctuation and special characters to focus on the core textual content. This step is crucial as it helps standardize the input and reduce noise in the data.

- splitting data in to train and validation set.

In [1]:
#import libraries
import re
import pandas as pd
import random
import torch
from datasets import load_dataset
from transformers import GPT2LMHeadModel, GPT2Tokenizer, Trainer, TrainingArguments, DataCollatorForLanguageModeling
from datasets import Dataset


'NoneType' object has no attribute 'cadam32bit_grad_fp32'


/Users/kumarkishalaya/anaconda3/lib/python3.11/site-packages/bitsandbytes/cextension.py:34: UserWarning: The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers, 8-bit multiplication, and GPU quantization are unavailable.
  warn("The installed version of bitsandbytes was compiled without GPU support. "


### Preparing data

In [2]:
data = pd.read_csv('df_commentary_new.csv')
data.drop('Unnamed: 0', axis=1, inplace=True)

In [3]:
data.head(5)

Modified_Commentary
0  FOUR, first boundary for batsman and team. Ful...
1  FOUR, back-to-back boundaries to end the first...
2  FOUR, hit straight back at bowler and he was l...
3  FOUR, another full toss, it's Jadhav this time...
4  FOUR, four more. Jadhav starting to really fin...

Splitting the dataset into training and validation sets

For fine-tuning, we need separate datasets for training and evaluation.

In [153]:
df_train_text= data.head(1000)
df_val_text = data.tail(1000)

In [155]:
# Function to remove punctuations and special characters
def clean_text(text):
    return re.sub(r'[^a-zA-Z0-9\s]', '', text)


In [156]:
# validation text
val_text = df_val_text['Modified_Commentary'].apply(clean_text).tolist()

In [157]:
val_text[:2]

['FOUR and Saini errs in length He goes full and batsman is just too good there extension of the arms and lofts it over extra cover the ball has enough to trickle into the fence',
 'out Caught by de Villiers Oh no batsman has nailed this but straight to the fielder Dubes golden arm strikes again Big call to trust him with this over after Saini had gone for a few but batsman has just thrown it away Charges down converts it into an overpitched delivery and drills it straight to AB at midoff The idea was to place it wide of the fielder like he done to Umesh but batsman was not able to do that batsman c de Villiers b bowler 1718 4s1']

In [160]:
# training text
train_text = df_train_text['Modified_Commentary'].apply(clean_text).tolist()

In [161]:
train_text[0:2]

['FOUR first boundary for batsman and team Full and on the pads needed to be put away and batsman did just that picked it up and dispatched it over midwicket couple of bounces and into the fence',
 'FOUR backtoback boundaries to end the first over Again bowler is a tad short in his length batsman had the width to cut and he didnt try to hit it hard just placed it behind point and Bhuvi at third man gave up the chase pretty quickly']

## Introduction to GPT-2

GPT-2 (Generative Pre-trained Transformer 2) is a large-scale language model developed by OpenAI. It's based on the transformer architecture, which utilizes self-attention mechanisms to process sequential data efficiently. Here are some key details about GPT-2:


### Architecture:

GPT-2 is a decoder-only transformer model, focusing on generating text based on previous tokens.
It uses masked self-attention layers, allowing each token to attend to all previous tokens in the sequence.
The model employs layer normalization and residual connections to facilitate training of deep networks.

<img src="/Users/kumarkishalaya/Desktop/Next_word_prediction/images/Screenshot 2024-07-06 at 6.51.35 AM.png" alt="Screenshot" width="600"/>

### Pre-training:

GPT-2 was pre-trained on a diverse dataset called WebText, containing about 8 million web pages.
The pre-training objective was next-token prediction, similar to what we're doing in our fine-tuning task.
This unsupervised pre-training allows the model to learn general language understanding and generation capabilities.



In [20]:
# Load tokenizer and model
model_name = 'gpt2'
tokenizer = GPT2Tokenizer.from_pretrained(model_name)
model = GPT2LMHeadModel.from_pretrained(model_name)
device = "cuda" if torch.cuda.is_available() else "cpu"
model.to(device)

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D()
          (c_proj): Conv1D()
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D()
          (c_proj): Conv1D()
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

In [163]:
# Create a dataset from the text list
dataset = Dataset.from_dict({"text": train_text})

#### Preparing Tokenizer to tokenize the data for the model

Tokenization is a critical step in preparing our data for the model. It converts the raw text into a format the model can process.

Here, we set a **maximum length of 512 tokens** and used padding to ensure all sequences are the same length. This standardization is necessary for batch processing during training. We could experiment with different max_length values to balance between capturing longer contexts and computational efficiency.


In [165]:
# Adding a padding token ensures that all sequences in a batch are of the same length.
tokenizer.pad_token = tokenizer.eos_token


# Tokenize the dataset
def tokenize_function(examples):
    return tokenizer(examples['text'], padding='max_length', truncation=True, max_length=512)

tokenized_datasets = dataset.map(tokenize_function, batched=True, remove_columns=['text'])

Map:   0%|          | 0/1000 [00:00<?, ? examples/s]

In [177]:
tokenized_datasets

Dataset({
    features: ['input_ids', 'attention_mask'],
    num_rows: 1000
})

#### Create data collator for next word prediction


**By setting mlm=False,** we specify that we're doing causal language modeling (next word prediction) rather than masked language modeling. This collator prepares the data in the format required for this specific task.

In [178]:
# keep mlm false to train on next word prediction
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)


#### Building Trainer Arguments: Setting up training parameters

Trainer arguments define how the model training process should be conducted.

This includes settings like **learning rate, batch size, and number of epochs,** which are crucial for effective training.Proper configuration of these parameters can significantly impact the model's performance and convergence.



In [ ]:
# Define training arguments
training_args = TrainingArguments(
    output_dir='./results',
    overwrite_output_dir=True,
    num_train_epochs=3,
    per_device_train_batch_size=2,
    save_steps=10_000,
    save_total_limit=2,
    logging_steps=20,
)

#### Training Step

In [49]:
# Create Trainer instance
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets,
    data_collator=data_collator,
)

# Train the model
trainer.train()

TrainOutput(global_step=1500, training_loss=1.8520173454284667, metrics={'train_runtime': 901.0243, 'train_samples_per_second': 3.33, 'train_steps_per_second': 1.665, 'total_flos': 783876096000000.0, 'train_loss': 1.8520173454284667, 'epoch': 3.0})

#### Save the model and tokenizer

In [72]:
model_save_path = 'code/saved_finetuned_gpt2_model'
trainer.save_model(model_save_path)
tokenizer.save_pretrained(model_save_path)

('./finetuned_gpt2/tokenizer_config.json',
 './finetuned_gpt2/special_tokens_map.json',
 './finetuned_gpt2/vocab.json',
 './finetuned_gpt2/merges.txt',
 './finetuned_gpt2/added_tokens.json')

#### Push model to huggingface hub

In [182]:
from huggingface_hub import login
login(token="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")

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: write).
Your token has been saved to /Users/kumarkishalaya/.cache/huggingface/token
Login successful


In [183]:
# Push the model to the hub
model.push_to_hub("Kumarkishalaya/GPT-2-next-word-prediction")
tokenizer.push_to_hub("Kumarkishalaya/GPT-2-next-word-prediction")

model.safetensors:   0%|          | 0.00/498M [00:00<?, ?B/s]

CommitInfo(commit_url='https://huggingface.co/Kumarkishalaya/GPT-2-next-word-prediction/commit/7246f3ecc6e18ad9c2ca85530f61d54810fabccc', commit_message='Upload tokenizer', commit_description='', oid='7246f3ecc6e18ad9c2ca85530f61d54810fabccc', pr_url=None, pr_revision=None, pr_num=None)

#### Loading models again for evaluation

In [184]:
model_load_path = 'code/saved_finetuned_gpt2_model'
model_finetuned = GPT2LMHeadModel.from_pretrained(model_load_path)
tokenizer_finetuned = GPT2Tokenizer.from_pretrained(model_load_path)

model_name = 'gpt2'
model_before_finetuning = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer_before_finetuning = GPT2Tokenizer.from_pretrained(model_load_path)

### Evaluation using perplexity and comparing a finetuned and a base model

In [185]:
# Function to calculate perplexity
def calculate_perplexity(text, model, tokenizer):
    encodings = tokenizer(text, return_tensors='pt')
    max_length = model.config.n_positions
    stride = 512
    seq_len = encodings.input_ids.size(1)
    
    nlls = []
    prev_end_loc = 0
    for begin_loc in range(0, seq_len, stride):
        end_loc = min(begin_loc + max_length, seq_len)
        trg_len = end_loc - prev_end_loc  # may be different from stride on last iteration
        input_ids = encodings.input_ids[:, begin_loc:end_loc]
        target_ids = input_ids.clone()
        target_ids[:, :-trg_len] = -100
        
        with torch.no_grad():
            outputs = model(input_ids, labels=target_ids)
            
            # loss is calculated using CrossEntropyLoss which averages over valid labels
            # N.B. the model only calculates loss over trg_len - 1 labels, because it internally shifts the labels
            # to the left by 1.
            neg_log_likelihood = outputs.loss * trg_len
        
        nlls.append(neg_log_likelihood)
        
        prev_end_loc = end_loc
        if end_loc == seq_len:
            break
    
    ppl = torch.exp(torch.stack(nlls).sum() / end_loc)
    return ppl.item()

In [188]:
device = "cuda" if torch.cuda.is_available() else "cpu"
model_finetuned.to(device)
model_before_finetuning.to(device)

# Set the model to evaluation mode
model_finetuned.eval()
model_before_finetuning.eval()


GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D()
          (c_proj): Conv1D()
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D()
          (c_proj): Conv1D()
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

In [212]:
val_text[:2]

['FOUR and Saini errs in length He goes full and batsman is just too good there extension of the arms and lofts it over extra cover the ball has enough to trickle into the fence',
 'out Caught by de Villiers Oh no batsman has nailed this but straight to the fielder Dubes golden arm strikes again Big call to trust him with this over after Saini had gone for a few but batsman has just thrown it away Charges down converts it into an overpitched delivery and drills it straight to AB at midoff The idea was to place it wide of the fielder like he done to Umesh but batsman was not able to do that batsman c de Villiers b bowler 1718 4s1']

In [202]:
# Calculate perplexity for each text
perplexity_finetuned_train_data = [calculate_perplexity(text,model_finetuned,tokenizer_finetuned) for text in train_text]
perplexity_finetuned = [calculate_perplexity(text,model_finetuned,tokenizer_finetuned) for text in val_text]
perplexity_base = [calculate_perplexity(text,model_before_finetuning,tokenizer_before_finetuning) for text in val_text]

In [207]:
# Calculate and print average perplexity
avg_perplexity_train_data = np.mean(perplexity_finetuned_train_data)
print(f"Average perplexity on training_data: {avg_perplexity_train_data:.2f}")


avg_perplexity_finetuned = np.mean(perplexity_finetuned)
print(f"Average perplexity on evalulation data: {avg_perplexity_finetuned:.2f}")

avg_perplexity_base = np.mean(perplexity_base)
print(f"Average perplexity on evaluation data (basemodel): {avg_perplexity_base:.2f}")


Average perplexity on training_data: 5.20
Average perplexity on evalulation data: 223.62
Average perplexity on evaluation data (basemodel): 574.18



Perplexity measures how well the model predicts a sample. **Lower perplexity indicates better performance**. We calculate perplexity for three scenarios:

1. **Fine-tuned model on training data**: This shows how well the model has learned the training set. **The lowest perplexity here (5.20)** indicates the model has fit well to the training data.

2. **Fine-tuned model on evaluation data:** This measures how well the model generalizes to unseen data in the same domain. **The higher perplexity (223.62)** compared to the training data suggests some overfitting, but it's still an improvement over the base model.

3. **Base model on evaluation data:** This serves as a baseline, showing the performance before fine-tuning. **The highest perplexity here (574.18)** demonstrates that fine-tuning has indeed improved the model's performance on cricket commentary.

The significant improvement from the base model to the fine-tuned model on evaluation data shows the value of domain-specific fine-tuning. However, the gap between training and evaluation perplexity for the fine-tuned model suggests there's room for improvement in generalization, possibly through techniques like regularization or using a larger and more diverse training dataset.

This fine-tuning process demonstrates how we can adapt a powerful language model like GPT-2 to a specific domain. While we've focused on next word prediction, this fine-tuned model could potentially be used for various downstream tasks related to cricket commentary, such as generating full commentaries, summarizing match highlights, or even answering cricket-related questions.

#### Playing around with output

In [74]:
device = "cuda" if torch.cuda.is_available() else "cpu"
model_finetuned.to(device)

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D()
          (c_proj): Conv1D()
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D()
          (c_proj): Conv1D()
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

In [211]:
input_text = 'how about that for a shot'
input_ids = tokenizer_finetuned(input_text, return_tensors="pt")
input_ids = input_ids['input_ids'].to(device)
output = model_finetuned.generate(input_ids, max_length=50, num_beams=5, do_sample=True)
loaded_tokenizer.decode(output[0])

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


'how about that for a shot Just a hint of width outside off and batsman takes it on the back foot and sends it sailing over deep midwicket No need to run for shots like that Length delivery that sits up to be smacked batsman'

In [29]:
model_before_finetuning = GPT2LMHeadModel.from_pretrained(model_name)
model_before_finetuning.to(device)

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D()
          (c_proj): Conv1D()
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D()
          (c_proj): Conv1D()
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

In [209]:
input_text = 'how about that for a shot'
input_ids = tokenizer(input_text, return_tensors="pt")
device = "cuda" if torch.cuda.is_available() else "cpu"
input_ids = input_ids['input_ids'].to(device)
output = model_before_finetuning.generate(input_ids, max_length=40, num_beams=5, do_sample=True)
tokenizer.decode(output[0])

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


'how about that for a shot?"\n\n"I don\'t think so."\n\n"I don\'t think so at all."\n\n"You don\'t think so?"\n\n"I'